components/builder-scheduler/src/data_store.rs

```diff
@@ -12,215 +12,335 @@
+                             END
+                          $$ LANGUAGE plpgsql VOLATILE"#)?;
+
+        migrator.migrate("scheduler",
+                         r#"CREATE INDEX pending_groups_index_v1 on groups(created_at) WHERE group_state = 'Pending'"#)?;
+
+        // Retrieve a group project
+        migrator.migrate("scheduler",
+                  r#"CREATE OR REPLACE FUNCTION find_project_v1 (gid bigint, name text) RETURNS SETOF projects AS $$
+                         BEGIN
+                           RETURN QUERY SELECT * FROM projects WHERE owner_id = gid AND project_name = name;
+                           RETURN;
+                         END
+                         $$ LANGUAGE plpgsql STABLE"#)?;
+
         Ok(())
     }
 
-    pub fn create_group(&mut self, project_names: Vec<String>) -> Result<Group> {
-        let mut projects = Vec::new();
-        for name in project_names {
-            let project = ProjectData {
-                name: name,
-                state: proto::ProjectState::NotStarted,
-            };
-            projects.push(project);
-        }
+    pub fn create_group(&self, project_names: Vec<String>) -> Result<Group> {
+        let conn = self.pool.get()?;
+
+        assert!(!project_names.is_empty());
 
+        // BUG - the insert query should be creating and assigning back a group_id,
+        // instead of expecting it to be passed in. The random id is a temporary
+        // workaround.
         let mut rng = thread_rng();
         let id = rng.gen::<u64>();
 
-        let group_data = GroupData {
-            id: id,
-            state: proto::GroupState::Pending,
-            projects: projects,
-            jobs: Vec::new(),
-        };
+        conn.execute("SELECT insert_group_v1($1, $2)",
+                     &[&(id as i64), &project_names])
+            .map_err(Error::GroupCreate)?;
+
 
-        self.groups.push(group_data.clone());
+        let mut projects = RepeatedField::new();
+
+        for name in project_names {
+            let mut project = Project::new();
+            project.set_name(name);
+            project.set_state(ProjectState::NotStarted);
+            projects.push(project);
+        }
 
-        assert!(!self.group_map.contains_key(&id));
-        self.group_map.insert(id, self.groups.len() - 1);
+        let mut group = Group::new();
+        group.set_id(id);
+        group.set_state(GroupState::Pending);
+        group.set_projects(projects);
 
-        println!("Group created: {:?}", group_data);
+        debug!("Group created: {:?}", group);
 
-        Ok(DataStore::group_data_to_group(&group_data))
+        Ok(group)
     }
 
-    pub fn get_group(&self, group_id: u64) -> Option<Group> {
-        if self.group_map.contains_key(&group_id) {
-            let index = *self.group_map.get(&group_id).unwrap();
-            assert!(self.groups[index].id == group_id);
+    pub fn get_group(&self, group_id: u64) -> Result<Option<Group>> {
```

Comment:
This function gets called from both a ScheduleGet request and also by a Job notification. So we could have it take a ScheduleGet as the incoming param, but that is one extra (though minor) step for the Job notification path.